feat: use typescript

SgtPooki opened the pull request:
See the gh-pages gateway checker website from my fork: https://sgtpooki.github.io/public-gateway-checker/

There are a number of changes included in this PR:

1. Migrated to typescript
2. Uses [ipfs/aegir](https://github.com/ipfs/aegir)
3. Static site is now in /dist, and README.md is updated to reflect that

lidel commented on package.json:
```diff
@@ -0,0 +1,154 @@
+    "copy-assets": "cp-cli src/index.html dist/index.html && cp-cli src/styles.css dist/styles.css",
+    "lint": "aegir lint && aegir ts -p check",
+    "release": "aegir release",
+    "publish-to-ipfs": "jsipfs add -wrQ dist > lastpubver",
```
This project uses GH Pages for now, this is not used, please remove `lastpubver` too
```suggestion
```

lidel reviewed (changes requested):
Thanks for cleaning this up @SgtPooki !

Did a quick pass and noticed some test regressions:

|  Old | This PR |
| ---- | ---- |
|   ![2022-03-03_17-21](https://user-images.githubusercontent.com/157609/156606581-d2eeeac9-ba63-47c3-8383-270b0af13e73.png)    |      ![2022-03-03_17-21_1](https://user-images.githubusercontent.com/157609/156606617-d808650e-1dc2-43c5-bf39-9b79e1dd2012.png) | 


Online column has false-negatives  and CORS test does not work at all – is this wip?

SgtPooki commented:
> Did a quick pass and noticed some test regressions
> Online column has false-negatives and CORS test does not work at all – is this wip?

I will investigate and update the PR

SgtPooki commented:
I noticed a few regressions:

1. "X/X Tested" string appears instantly with changes from this PR. It should only count gateways as they're actually tested.
2. Cors field "onerror" should be empty on error (@lidel do we want to keep this logic... or should we make the UI consistent by showing X icon or caution?)
3. Some pending icons (clocks) show forever.

SgtPooki commented:
@lidel I was able to serialize the results of the site [before my changes](https://github.com/ipfs/public-gateway-checker/pull/194/commits/3209b76ef6ada27fd49f6d92f0e6adb42749c277), and [after](https://github.com/ipfs/public-gateway-checker/pull/194/commits/d7b05973155c6bf3dd6372e475f3170887679e0e)

I believe the difference lies in the fault of the existing logic, and that I've removed all obvious regressions except for one: The total gateways tested still shows "90/90 tested" on page load, immediately. I'll look into that now, but let me know what you think about the rest.

I also significantly improved the DNS lookup (more intelligent rate-limiting with an additional DNS-query endpoint via google) and the timings of the various asynchronous things. I also replaced two XMLHttpRequest calls with fetch calls. There is only one left, in the `expectSubdomainRedirect` function.

lidel commented on .github/workflows/build-and-publish-github-pages.yml:
```diff
@@ -0,0 +1,26 @@
+
+      # Deploy to gh pages branch
+      - name: Deploy
+        uses: s0/git-publish-subdir-action@develop
```
security precaution
```suggestion
        uses: s0/git-publish-subdir-action@399aab378450f99b7de6767f62b0d1dbfcb58b53
```

lidel commented on package.json:
```diff
@@ -0,0 +1,154 @@
+{
+  "name": "@ipfs/public-gateway-checker",
```
Ensure we don't publish to NPM by mistake :))
```suggestion
  "name": "@ipfs/public-gateway-checker",
  "private": true,
```

lidel commented on .github/workflows/js-test-and-release.yml:
```diff
@@ -0,0 +1,131 @@
+name: test & maybe release
```
I don't think we have any tests, should we comment-out `test-*` jobs below?

lidel commented on .github/workflows/js-test-and-release.yml:
```diff
@@ -0,0 +1,131 @@
+    strategy:
+      matrix:
+        os: [windows-latest, ubuntu-latest, macos-latest]
+        node: [16]
```
Since we use lts everywhere else:
```suggestion
        node: ['lts/*']
```

lidel commented on package.json:
```diff
@@ -0,0 +1,154 @@
+    ],
+    "plugins": [
+      [
+        "@semantic-release/commit-analyzer",
```
Q:  are we using `@semantic-release` anywhere?

SgtPooki replied:
nope. Not yet. I believe this was added by `check-aegir-project` I can remove this, but we may want to release this as an npm package eventually. See https://github.com/ipfs/public-gateway-checker/issues/192

SgtPooki replied:
closing this in favor of eventually releasing as npm package, #192

lidel reviewed (changes requested):
@SgtPooki LGTM, thank you for cleaning this all up! :heart: 

Before we merge this, mind addressing small nits described below?

SgtPooki commented on package.json:
```diff
@@ -0,0 +1,154 @@
+    "test-TBD:browser": "aegir test --target browser"
+  },
+  "dependencies": {
+    "@dutu/rate-limiter": "github:dutu/rate-limiter",
```
let me look into this

SgtPooki replied:
I'm going to stick with rate-limiter as it seems significantly less complicated to use while allowing more control due to the implemented token-bucket algorithm. let me know if you feel strongly about this

SgtPooki commented on src/index.html:
```diff
@@ -50,15 +53,16 @@ <h1 class='f3 fw2 montserrat aqua ttu ma0'>Public Gateways</h1>
 		</div>
 	</div>
-<script src="./app.js?v=0.4"></script>
+<script src="./index.min.js?v=0.4"></script>
```
Moved to an issue to address later: https://github.com/ipfs/public-gateway-checker/issues/195

SgtPooki commented:
Good to merge now @lidel ?
